ADApp/pluginSrc/NDFileHDF5AttributeDataset.cpp

```diff
@@ -185,36 +185,61 @@ asynStatus NDFileHDF5AttributeDataset::writeAttributeDataset(hdf5::When_t whenTo
       extendIndexDataSet(offsets[indexed]);
     }
     // find the data based on datatype
-    ret = ndAttr->getValue(ndAttr->getDataType(), pDatavalue, MAX_ATTRIBUTE_STRING_SIZE);
+    ret = ndAttr->getValue(ndAttr->getDataType(), pDatavaluestore[dataset_count], MAX_ATTRIBUTE_STRING_SIZE);
     if (ret == ND_ERROR) {
-      memset(pDatavalue, 0, MAX_ATTRIBUTE_STRING_SIZE);
+      memset(pDatavaluestore[dataset_count], 0, MAX_ATTRIBUTE_STRING_SIZE);
+    }
+    dataset_count++;
+    // Check if we are being asked to flush or if the size of the store has reached its maximum amount
+    if (flush == 1 or dataset_count == MAX_BATCH_SIZE){
+      elementSize_[0] = dataset_count;
+      // Offset has been incremented, so is at where the stored data should end
+      offset_[0] -= (dataset_count - 1);
+      // Work with HDF5 library to select a suitable hyperslab (one element) and write the new data to it
+      H5Dset_extent(dataset_, dims_);
+      filespace_ = H5Dget_space(dataset_);
+      memspace_ = H5Screate_simple(rank_, elementSize_, NULL);
+      // Select the hyperslab
+      H5Sselect_hyperslab(filespace_, H5S_SELECT_SET, offset_, NULL, elementSize_, NULL);
+      H5Dwrite(dataset_, datatype_, memspace_, filespace_, H5P_DEFAULT, pDatavaluestore);
+      if (flush ==1)
+        status = this->flushDataset();
+      H5Sclose(filespace_);
+      dataset_count = 0;
     }
+
+    nextRecord_++;
+  }
+
+  return status;
+}
```

Comment:
The behavior here has been changed to not match the other overload. Is there any reason not to also change the other overload to only call do the write if flush == 1 or MAX_BATCH_SIZE is reached?

Reply:
As far as I can tell, the only difference between the two existing methods is whether the datasets should match the dimensions of the detector data, or just be 1D. It seems to me that this new functionality should be added in both places, as you might want the data to be written fast in 1D. @ajgdls is this correct?

Again this raises the question of whether the duplicated code in these methods could be reduced.

Reply:
From my testing, I had noticed that the other method only gets called at the start of the scan, once for each dataset and it doesn't actually get to the H5Dwrite call, so batching the writes wouldn't have made any difference.

I should be able to move my changes across to the other method, but as you say it would duplicate code. I could call my new method from the existing two for the actual write, and keep the existing methods' preambles as that's where the current differences are?

Reply:
I've updated my changes, having taken the common code for writing the datasets into it's own private method which the existing methods call. Do you think that this is a better implementation?

Reply:
I think it is better.